compiler/src/iree/compiler/Codegen/LLVMCPU/DispatchABI.cpp

```diff
@@ -785,12 +793,31 @@ MemRefDescriptor HALDispatchABI::loadBinding(Operation *forOp, int64_t ordinal,
     // Compute and update strides. Assume that MemRefs are row-major, that is,
     // following index linearization:
     //   x[i, j, k] = i * x.dim[1] * x.dim[2] + j * x.dim[2] + k
-    desc.setConstantStride(builder, loc, rank - 1, 1);
-    for (int i = rank - 2; i >= 0; --i) {
-      auto stride = desc.stride(builder, loc, i + 1);
-      auto dim = desc.size(builder, loc, i + 1);
-      Value strideVal = builder.create<LLVM::MulOp>(loc, stride, dim);
-      desc.setStride(builder, loc, i, strideVal);
+    if (!strides.empty()) {
+      assert(strides.back() == 1 &&
+             "unexpected non-unit stride for innermost dimension");
+      desc.setConstantStride(builder, loc, rank - 1, 1);
+      OpFoldResult currentStride = builder.getIndexAttr(1);
+      for (int i = rank - 1; i > 0; --i) {
```

Comment:
This similarly looks like upstream logic you don't want to be duplicating and inlining at every use place in IREE.
I can't point offhand to the right API but such functionality should be exposed here independently: mlir/include/mlir/Conversion/LLVMCommon/MemRefBuilder.h

Reply:
I disagree, this is not small sharing but a tricky API that is easy to get wrong and that must not proliferate: every code path then gets its own, someday someone refactors/fixes/introduces a bug and now we have different implementations, each with its own quirks..

The answer is to design good/reusable APIs.
My recommendation is you start something like mlir/include/mlir/Dialect/Utils/IndexingUtils.h but in a place that can depend on LLVM (we prob need the same for creating AffineApplyOps with values in a similar place that can depend on Affine).

Having consistent APIs across the board will help reduce than chance for footguns.
I acknowledge this is work but it is important.

Reply:
Yes I am fine with that, let's put a high pri on this if you don't mind, this type of code has a tendency of being rewritten many times over the codebase.

I'd suggest `mlir/include/mlir/Dialect/LLVM/Utils/IndexingUtils.h` as a starting point.